### PR TITLE
feat: ephemeral notebook

### PR DESCRIPTION
**Pull request by drdelambre**
base mechanism for ephemeral notebook. there are still some things to do based on the figma, but i wanna make sure people take it for a spin before we start investing more into it.

Peek 2021-09-23 14-21

### CODE REVIEW

**Comment by ui-ci-bot-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 62 | :yellow_circle:
accessibility | 96 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :hankey:
best-practices | 100 | :green_circle:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?lower=now%28%29%20-%202d&vars%5Blighthouse_score_branch%5D=2690)

File: src/flows/context/flow.current.tsx [hunk 1]
```diff
@@ -32,43 +33,86 @@ let GENERATOR_INDEX = 0
 
 export const FlowProvider: FC = ({children}) => {
   const {flows, update, currentID} = useContext(FlowListContext)
+  const [currentFlow, setCurrentFlow] = useState<Flow>()
+
+  // NOTE this is a pretty aweful mecahnism, as it duplicates the source of
+  // thruth for the definition of the current flow, but i can't see a good
+  // way around it. We need to ensure that we're still updating the values
+  // and references to the flows object directly to get around the async
+  // update issues.
```
Comment (asalem1) [inline src/flows/context/flow.current.tsx:42]:
```suggestion
  // NOTE this is a pretty awful mechanism, as it duplicates the source of
  // truth for the definition of the current flow, but i can't see a good
  // way around it. We need to ensure that we're still updating the values
  // and references to the flows object directly to get around the async
  // update issues.
```

**Review by asalem1 — APPROVED** [review]
Honestly looks solid. A couple of things that shouldn't hold up the PR but I know could be helpful would just be to:

1. Maybe make a comment or issue as to why some of the conditional logic is necessary (i.e. we don't wanna update the API). I know why it's there, but new hire X won't and I'll look back in a year and think "what the heck Boatie"
2. Is it worth following up on the comment you wrote about not liking what had to be done in the useEffect to possible address it some other way later on?
3. rename the feature flag to `ephemeralNotebook`. I couldn't care less about the name, but there are people who will

File: src/flows/context/flow.current.tsx [hunk 2]
```diff
@@ -32,43 +33,86 @@ let GENERATOR_INDEX = 0
...
+        setCurrentFlow({...currentFlow})
+        return
+      }
+
       flows[currentID].data.byID[id] = {
         ...(flows[currentID].data.byID[id] || {}),
         ...data,
       }
 
+      // this should update the useEffect on the next time around
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/flows/context/flow.current.tsx:65]: it's a signpost to signal that we're not supposed to be relying on the useState as the source of truth for notebooks with an id, as this'll overwrite it. intentionally.

File: src/flows/context/flow.current.tsx [hunk 3]
```diff
@@ -32,43 +33,86 @@ let GENERATOR_INDEX = 0
...
     (id: string, meta: Partial<PipeMeta>) => {
+      if (isFlagEnabled('ephemeral') && !currentFlow.id) {
+        currentFlow.meta.byID[id] = {
+          title: '',
+          visible: true,
+          ...(currentFlow.meta.byID[id] || {}),
+          ...meta,
+        }
+        setCurrentFlow({...currentFlow})
+        return
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/flows/context/flow.current.tsx:85]: this just says "if you're in ephemeral mode, use the local `useState` state. if you're not, use the list state"

File: src/flows/context/flow.current.tsx [hunk 4]
```diff
@@ -32,43 +33,86 @@ let GENERATOR_INDEX = 0
 
 export const FlowProvider: FC = ({children}) => {
   const {flows, update, currentID} = useContext(FlowListContext)
+  const [currentFlow, setCurrentFlow] = useState<Flow>()
+
+  // NOTE this is a pretty awful mechanism, as it duplicates the source of
+  // truth for the definition of the current flow, but i can't see a good
+  // way around it. We need to ensure that we're still updating the values
+  // and references to the flows object directly to get around the async
+  // update issues.
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/flows/context/flow.current.tsx:42]: i would have just fixed it, but i think there's something inherent in supporting both paradigms with the flag that is preventing the normalization and is going to push it off until we remove the flag. at which point, i hope someone finds this comment

File: src/flows/components/header/SaveState.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    add(flow).then(id => {
+      history.replace(`/orgs/${org.id}/notebooks/${id}`)
+    })
+  }
+
+  if (!isFlagEnabled('ephemeral')) {
+    return null
+  }
+
+  if (!flow.id) {
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/flows/components/header/SaveState.tsx:33]: it'll never reach this condition

File: src/flows/components/FromTemplatePage.tsx [hunk 6]
```diff
@@ -33,17 +49,51 @@ const Template: FC = () => {
...
+              <Switch>
+                <Route path="/notebook/from/*" component={Template} />
+                <Route component={NotFound} />
+              </Switch>
+            </FlowProvider>
+          </FlowListProvider>
+        </QueryProvider>
+      </AppWrapper>
+    )
+  }
```
[the remark this answers is not in the document]
Comment (drdelambre) [reply src/flows/components/FromTemplatePage.tsx:95]: omg tell me about it. gotta suppress a lot of internal anguish about this one

File: src/flows/context/flow.current.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (asalem1) [reply src/flows/context/flow.current.tsx:85]: So if that's the case, will we have a conflict when the ephemeral notebook is saved?

[the remark this answers is not in the document]
Comment (drdelambre) [reply src/flows/context/flow.current.tsx:85]: Nope. Cause then you're not using an ephemeral notebook, just a normal one

**Pull request merged**